packages/flutter_tools/lib/src/build_system/targets/darwin.dart

```diff
@@ -0,0 +1,147 @@
+
+  // Check that dev dependencies are enabled on debug builds and disabled on
+  // release or profile builds.
+  //
+  // The Flutter tool enables/disables dev dependencies using the build mode.
+  // These can get out of sync if the user switches the build mode in Xcode.
+  @override
+  Future<void> build(Environment environment) async {
+    final String? buildModeEnvironment = environment.defines[kBuildMode];
+    if (buildModeEnvironment == null) {
+      throw MissingDefineException(kBuildMode, name);
+    }
+    final BuildMode buildMode = BuildMode.fromCliName(buildModeEnvironment);
+    final String? devDependenciesEnabledString = environment.defines[kDevDependenciesEnabled];
+    if (devDependenciesEnabledString == null) {
+      throw MissingDefineException(kDevDependenciesEnabled, name);
+    }
+
+    final bool? devDependenciesEnabled = bool.tryParse(
+      environment.defines[kDevDependenciesEnabled] ?? '',
+    );
+    if (devDependenciesEnabled == null) {
+      throw Exception(
+        'Unexpected $kDevDependenciesEnabled define value: "$devDependenciesEnabledString"',
+      );
+    }
+
+    if (devDependenciesEnabled && buildMode.isRelease) {
+      // Supress this error if the project has no dev dependencies.
+      if (!_hasDevDependencies(environment)) {
+        environment.logger.printTrace(
+          'Ignoring dev dependencies error as the project has no dev dependencies',
+        );
+        return;
+      }
+
+      _printXcodeError(
+        'Release builds should not have Dart dev dependencies enabled\n'
+        '\n'
+        'This error happens if:\n'
+        '\n'
+        '1. Your pubspec.yaml has dev dependencies\n'
+        '2. Your last Flutter CLI action turned on dev dependencies\n'
+        '3. You do a release build in Xcode\n'
+        '\n'
+        'You can turn off Dart dev dependencies by running this in your Flutter project:\n'
+        '\n'
+        '  $releaseBuildCommand\n'
+        '\n',
+      );
+      throwToolExit('Dev dependencies enabled in release build');
+    } else if (!devDependenciesEnabled && !buildMode.isRelease) {
+      // Supress this error if the project has no dev dependencies.
+      if (!_hasDevDependencies(environment)) {
+        environment.logger.printTrace(
+          'Ignoring dev dependencies error as the project has no dev dependencies',
+        );
+        return;
+      }
+
+      final bool profile = buildMode == BuildMode.profile;
+      _printXcodeError(
+        '${profile ? 'Profile' : 'Debug'} builds require Dart dev dependencies\n'
+        '\n'
+        'This error happens if:\n'
+        '\n'
+        '1. Your pubspec.yaml has dev dependencies\n'
+        '2. Your last Flutter CLI action turned off dev dependencies\n'
+        '3. You do a debug or profile build in Xcode\n'
+        '\n'
+        'You can turn on Dart dev dependencies by running this in your Flutter project:\n'
+        '\n'
+        '  $debugBuildCommand\n'
```

Comment:
Perhaps should calculate this field rather than pass it in? You should probably do `flutter build ios --config-only --profile` for profile mode, not `--debug`. It's weird to do `--debug` for a `--profile` build

Reply:
Updated to show `--profile` for profile mode!